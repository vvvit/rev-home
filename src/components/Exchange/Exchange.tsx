import * as React from 'react';
import {cn} from '@bem-react/classname';

import { CurrencyCard, CurrencyCardType } from '../CurrencyCard/CurrencyCard';

const cnApp = cn('Exchange');

class Exchange extends React.Component {
  render() {
    return (
      <div className={cnApp()}>
        <CurrencyCard
            currency="RUR"
            type={CurrencyCardType.buy}
        />
        <CurrencyCard
            currency="USD"
            type={CurrencyCardType.sell}
        />
      </div>
    );
  }
}

export { Exchange };
