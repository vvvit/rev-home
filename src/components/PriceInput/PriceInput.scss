@import '../../styles/variables.scss';

.PriceInput {
    &-Input {
        width: 100%;
        font-size: $font-size-xl;
        text-align: right;
        outline: 0;
        border: 0;
        box-sizing: border-box;
        background: transparent;
    }
}
