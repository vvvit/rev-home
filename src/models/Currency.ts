export const enum CurrencyName {
    USD = 'usd',
    EUR = 'eur',
    GBP = 'gbp',
}

export interface ICurrency {
    name: CurrencyName,
    value: number
}
