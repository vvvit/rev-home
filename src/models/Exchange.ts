export const enum CurrencyExchangeType {
    BUY = 'buy',
    SELL = 'sell'
}
