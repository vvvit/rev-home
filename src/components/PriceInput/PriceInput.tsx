import * as React from 'react';
import {boundMethod} from 'autobind-decorator';
import {cn} from "@bem-react/classname";

import './PriceInput.scss';

interface IPriceInputProps {
    value: string;
    onChange: (value: number) => void;
}

interface IPriceInputState {
    value: string;
}

const cnPriceInput = cn('PriceInput');

export const MAX_VALUE_LENGHT = 15;

export class PriceInput extends React.Component<IPriceInputProps, IPriceInputState> {
    static defaultProps: Partial<IPriceInputProps> = {
        onChange: () => {},
        value: '0'
    };

    state: IPriceInputState = {
        value: this.props.value
    };

    @boundMethod
    onChange(e: React.FormEvent<HTMLInputElement>) {
       e.persist();

        const value = this.formatValue(e.currentTarget.value);

        this.setState({ value });
    }

    render() {
        return (
            <div className={cnPriceInput()}>
                <input className={cnPriceInput('Input')} value={this.state.value} onChange={this.onChange} />
            </div>
        );
    }

    formatValue(value: string) {
        // valid float string without insignificant trailing zeros
        const validNumber = parseFloat(parseFloat(value).toFixed(2));
        let validString = isNaN(validNumber) ? '0' : String(validNumber);

        if (value.length > MAX_VALUE_LENGHT) {
            validString = value.substring(-value.length + MAX_VALUE_LENGHT, MAX_VALUE_LENGHT);
        } else if (validString.length < (MAX_VALUE_LENGHT - 2) && value.indexOf('.0') === validString.length) {
            // for strings like '11.0'
            validString += '.0';
        } else if (validString.length < (MAX_VALUE_LENGHT - 1) && value.indexOf('.') === validString.length) {
            // for strings like '11.'
            validString += '.';
        }

        return validString;
    }
}
