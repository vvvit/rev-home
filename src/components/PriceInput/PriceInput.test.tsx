import * as React from 'react';
import { shallow } from 'enzyme';

import { PriceInput, MAX_VALUE_LENGHT } from './PriceInput';
import { CurrencyExchangeType } from '../../models/Exchange';

describe('PriceInput', () => {
    let onChange: jest.Mock;
    const event = { persist: jest.fn(), currentTarget: { value: '1' }};

    beforeEach(() => {
        onChange = jest.fn();
        event.currentTarget.value = '0';
    });

    it('renders correctly with BUY type', () => {
        expect(shallow(
            <PriceInput
                value="100.01"
                type={CurrencyExchangeType.BUY}
                onChange={onChange}
            />
        )).toMatchSnapshot();
    });

    it('renders correctly with SELL type', () => {
        expect(shallow(
            <PriceInput
                value="990.01"
                type={CurrencyExchangeType.SELL}
                onChange={onChange}
            />
        )).toMatchSnapshot();
    });

    it('call `onChange` when the value is changed', () => {
        const component = shallow(
            <PriceInput
                value="990.01"
                type={CurrencyExchangeType.SELL}
                onChange={onChange}
            />
        );

        event.currentTarget.value = '1';
        component.find('.PriceInput-Input')
            .simulate('change', event);
        expect(onChange.mock.calls[0][0]).toBe('1');

        event.currentTarget.value = '- 100.01';
        component.find('.PriceInput-Input')
            .simulate('change', event);
        expect(onChange.mock.calls[1][0]).toBe('100.01');
    });

    describe('.formatValue', () => {
        const wrapper = shallow<PriceInput>(
            <PriceInput/>
        );
        const priceInputInstace = wrapper.instance();

        it('returns integer string if receives integer string', () => {
            expect(priceInputInstace.formatValue('111')).toEqual('111');
            expect(priceInputInstace.formatValue('1234')).toEqual('1234');
        });

        it('returns float string if receives float string', () => {
            expect(priceInputInstace.formatValue('111.1')).toEqual('111.1');
            expect(priceInputInstace.formatValue('1234.3')).toEqual('1234.3');
        });

        it('valid number can\'t be more than max length', () => {
            const value = new Array(MAX_VALUE_LENGHT + 1).join('1');

            expect(priceInputInstace.formatValue(`${value}2`)).toEqual(value);
        });

        it('string with not only digits formats to float', () => {
            expect(priceInputInstace.formatValue('33error')).toEqual('33');
            expect(priceInputInstace.formatValue('33.3error')).toEqual('33.3');
            expect(priceInputInstace.formatValue('33.3.3')).toEqual('33.3');
            expect(priceInputInstace.formatValue('3+')).toEqual('3');
        });

        it('returns \'0\' for string which can\'t be formed', () => {
            expect(priceInputInstace.formatValue('wrong')).toEqual('0');
            expect(priceInputInstace.formatValue('--33.3')).toEqual('0');
        });

        it('saves dot in the end for digits string', () => {
            expect(priceInputInstace.formatValue('12.')).toEqual('12.');
        });

        it('saves first zero after dot in the valid float string', () => {
            expect(priceInputInstace.formatValue('12.0')).toEqual('12.0');
        });

        it('erase dot if length with dot equal max length of the string', () => {
            const value = new Array(MAX_VALUE_LENGHT).join('1');

            expect(priceInputInstace.formatValue(`${value}.`)).toEqual(`${value}`);
        });

        it('erase last zero after dot if length equal max length of the string', () => {
            const value = new Array(MAX_VALUE_LENGHT - 1).join('1');

            expect(priceInputInstace.formatValue(`${value}.0`)).toEqual(`${value}.`);
        });
    });
});
