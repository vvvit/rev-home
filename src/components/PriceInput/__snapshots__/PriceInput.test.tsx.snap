// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceInput renders correctly with BUY type 1`] = `
<div
  className="PriceInput"
>
  <input
    className="PriceInput-Input"
    onChange={[Function]}
    onFocus={[Function]}
    value="+ 100.01"
  />
</div>
`;

exports[`PriceInput renders correctly with SELL type 1`] = `
<div
  className="PriceInput"
>
  <input
    className="PriceInput-Input"
    onChange={[Function]}
    onFocus={[Function]}
    value="- 990.01"
  />
</div>
`;
